extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,347 @@
+
+  /** @override */
+  report() {
+    this.form.reportValidity();
+  }
+
+}
+
+
+class PolyfillDefaultValidator extends FormValidator {
+
+  constructor(form) {
+    super(form);
+
+    if (!validationBubble) {
+      const win = this.doc.defaultView;
+      validationBubble = new ValidationBubble(win, 'amp-validation-bubble');
+    }
+  }
+
+  /** @override */
+  report() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      if (!inputs[i].checkValidity()) {
+        inputs[i]./*REVIEW*/focus();
+        validationBubble.show(inputs[i], inputs[i].validationMessage);
+        break;
+      }
+    }
+  }
+
+  /** @override */
+  onBlur(unusedEvent) {
+    validationBubble.hide();
+  }
+
+  /** @override */
+  onInput(event) {
+    if (event.target.checkValidity()) {
+      event.target.removeAttribute('aria-invalid');
+      validationBubble.hide();
+    } else {
+      event.target.setAttribute('aria-invalid', 'true');
+      validationBubble.show(event.target, event.target.validationMessage);
+    }
+  }
+}
+
+
+class AbstractCustomValidator extends FormValidator {
+
+  /**
+   * @param {!Element} input
+   */
+  reportInput(input) {
+    const invalidType = getInvalidType(input);
+    if (invalidType) {
+      this.showValidationFor(input, invalidType);
+    }
+  }
+
+  /**
+   * Hides all validation messages.
+   */
+  hideAllValidations() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+
+    return this.doc.querySelector(
+        `[visible-when-invalid=${invalidType}][validation-for=${input.id}]`);
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {string} invalidType
+   */
+  showValidationFor(input, invalidType) {
+    const validation = this.getValidationFor(input, invalidType);
+    if (!validation) {
+      return;
+    }
+
+    if (!validation.textContent.trim()) {
+      validation.textContent = input.validationMessage;
+    }
+
+    input.setAttribute('aria-invalid', 'true');
+    validation.classList.add('visible');
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {boolean} Whether a validation was visible and was successfully hidden.
+   */
+  hideValidationFor(input) {
+    const visibleValidation = this.getVisibleValidationFor(input);
+    if (!visibleValidation) {
+      return false;
+    }
+
+    input.removeAttribute('aria-invalid');
+    visibleValidation.classList.remove('visible');
+    return true;
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {?Element}
+   */
+  getVisibleValidationFor(input) {
+    if (!input.id) {
+      return null;
+    }
+    return this.doc.querySelector(
```

Comment:
Yeah. Would querying once the validator is instantiated and store all validation elements in instance variables and then looping over to check which is visible be a better solution? Or instead of looping we can keep track of visible validations all the time in an array or a dict instance variable.

What do you think?